src/data/bucket/fill_extrusion_bucket.js

```diff
@@ -359,142 +363,188 @@ class FillExtrusionBucket implements Bucket {
-                            // │ 2  3 │ Triangle 2: 1 => 2 => 3
-                            // └──────┘
-                            this.indexArray.emplaceBack(bottomRight, bottomRight + 2, bottomRight + 1);
-                            this.indexArray.emplaceBack(bottomRight + 1, bottomRight + 2, bottomRight + 3);
+                    const isConcaveCorner = ring.length > 4 && isAOConcaveAngle(p0, p1, p2);
+                    let encodedEdgeDistance = encodeAOToEdgeDistance(edgeDistance, isPrevCornerConcave, true);
 
-                            segment.vertexLength += 4;
-                            segment.primitiveLength += 2;
+                    addVertex(this.layoutVertexArray, p0.x, p0.y, nxRatio, nySign, 0, 0, encodedEdgeDistance);
+                    addVertex(this.layoutVertexArray, p0.x, p0.y, nxRatio, nySign, 0, 1, encodedEdgeDistance);
 
-                            if (isGlobe) {
-                                const array: any = this.layoutVertexExtArray;
+                    edgeDistance += dist;
+                    encodedEdgeDistance = encodeAOToEdgeDistance(edgeDistance, isConcaveCorner, false);
+                    isPrevCornerConcave = isConcaveCorner;
 
-                                const projectedP1 = projection.projectTilePoint(p1.x, p1.y, canonical);
-                                const projectedP2 = projection.projectTilePoint(p2.x, p2.y, canonical);
+                    addVertex(this.layoutVertexArray, p1.x, p1.y, nxRatio, nySign, 0, 0, encodedEdgeDistance);
+                    addVertex(this.layoutVertexArray, p1.x, p1.y, nxRatio, nySign, 0, 1, encodedEdgeDistance);
 
-                                const n1 = projection.upVector(canonical, p1.x, p1.y);
-                                const n2 = projection.upVector(canonical, p2.x, p2.y);
+                    segment.vertexLength += 4;
 
-                                addGlobeExtVertex(array, projectedP1, n1);
-                                addGlobeExtVertex(array, projectedP1, n1);
-                                addGlobeExtVertex(array, projectedP2, n2);
-                                addGlobeExtVertex(array, projectedP2, n2);
-                            }
-                        }
-                    }
-                }
-            }
+                    // ┌──────┐
+                    // │ 1  3 │ clockwise winding order.
+                    // │      │ Triangle 1: 0 => 1 => 2
+                    // │ 0  2 │ Triangle 2: 1 => 3 => 2
+                    // └──────┘
+                    this.indexArray.emplaceBack(k + 0, k + 1, k + 2);
+                    this.indexArray.emplaceBack(k + 1, k + 3, k + 2);
+                    segment.primitiveLength += 2;
 
-            if (segment.vertexLength + numVertices > SegmentVector.MAX_VERTEX_ARRAY_LENGTH) {
-                segment = this.segments.prepareSegment(numVertices, this.layoutVertexArray, this.indexArray);
-            }
+                    if (edgeRadius) {
```

Comment:
I really think it would be great if these parts have more code comments (or perhaps even an illustration in the PR description). The algorithm is arguably simple but at the moment reasoning about it is somewhat difficult.